File: injector/worker/http-worker/internal/model/task.go
package model

import (
	"encoding/json"
	"time"
)

// HttpTask 代表从Kafka接收的HTTP任务
type HttpTask struct {
	TaskID  string      `json:"taskId"`
	Payload TaskPayload `json:"payload"`
}

// TaskPayload 任务载荷
type TaskPayload struct {
	DataSourceURL string                 `json:"dataSourceUrl"`
	Method        string                 `json:"method"`
	Params        map[string]interface{} `json:"params"`
	APIKey        string                 `json:"apikey"`
	DataSourceID  string                 `json:"dataSourceId"`
	Headers       map[string]string      `json:"headers"`
}

// TaskStatus 任务状态
type TaskStatus struct {
	TaskID      string    `json:"taskId"`
	Status      string    `json:"status"`      // SUCCESS, FAILED, RETRY
	Message     string    `json:"message"`     // 错误信息或状态描述
	Timestamp   time.Time `json:"timestamp"`   // 处理时间戳
	Duration    int64     `json:"duration"`    // 处理耗时（毫秒）
	StatusCode  int       `json:"statusCode"`  // HTTP响应状态码
	DataSize    int       `json:"dataSize"`    // 响应数据大小
	RetryCount  int       `json:"retryCount"`  // 重试次数
}

// DataRecord 数据记录，用于发送到数据topic
type DataRecord struct {
	TaskID       string                 `json:"taskId"`
	DataSourceID string                 `json:"dataSourceId"`
	Timestamp    time.Time              `json:"timestamp"`
	Data         interface{}            `json:"data"`
	Metadata     map[string]interface{} `json:"metadata"`
}

// HTTPResponse HTTP响应结构
type HTTPResponse struct {
	StatusCode int               `json:"statusCode"`
	Headers    map[string]string `json:"headers"`
	Body       interface{}       `json:"body"`
	Duration   time.Duration     `json:"duration"`
	Size       int               `json:"size"`
}

// IsRetryable 判断状态码是否可重试
func (r *HTTPResponse) IsRetryable() bool {
	// 5xx错误和网络相关错误可重试
	if r.StatusCode >= 500 && r.StatusCode < 600 {
		return true
	}
	// 429限流错误可重试
	if r.StatusCode == 429 {
		return true
	}
	// 408请求超时可重试
	if r.StatusCode == 408 {
		return true
	}
	return false
}

// ToJSON 将对象转换为JSON字符串
func (t *HttpTask) ToJSON() (string, error) {
	data, err := json.Marshal(t)
	return string(data), err
}

// FromJSON 从JSON字符串解析HttpTask
func (t *HttpTask) FromJSON(jsonStr string) error {
	return json.Unmarshal([]byte(jsonStr), t)
}

// ToJSON 将TaskStatus转换为JSON字符串
func (ts *TaskStatus) ToJSON() (string, error) {
	data, err := json.Marshal(ts)
	return string(data), err
}

// ToJSON 将DataRecord转换为JSON字符串
func (dr *DataRecord) ToJSON() (string, error) {
	data, err := json.Marshal(dr)
	return string(data), err
}

File: datainjector/control-plane-service/src/main/java/com/crypto/control/model/Task.java
package com.crypto.control.model;

import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.Builder;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.Map;

/**
 * 任务实体类
 * 对应数据库中的tasks表
 */
@Entity
@Table(name = "tasks")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Task {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(name = "task_id", unique = true, nullable = false, length = 64)
    private String taskId;
    
    @Column(name = "data_source_id", nullable = false, length = 64)
    private String dataSourceId;
    
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 16)
    @Builder.Default
    private TaskStatus status = TaskStatus.PENDING;
    
    @Column(name = "retry_count", nullable = false)
    @Builder.Default
    private Integer retryCount = 0;
    
    @Column(name = "max_retry_count", nullable = false)
    @Builder.Default
    private Integer maxRetryCount = 3;

    @Column(name = "task_type", length = 64)
    private String taskType;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "payload", columnDefinition = "jsonb")
    private Map<String, Object> payload;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "metadata", columnDefinition = "jsonb")
    private Map<String, Object> metadata;
    
    @Column(name = "scheduled_time", nullable = false)
    private LocalDateTime scheduledTime;    
    @Column(name = "started_at")
    private LocalDateTime startedAt;
    
    @Column(name = "completed_at")
    private LocalDateTime completedAt;
    
    @Column(name = "status_code")
    private Integer statusCode;
    
    @Column(name = "message", columnDefinition = "TEXT")
    private String message;
    
    @Column(name = "duration_ms")
    private Long durationMs;
    
    @Column(name = "data_size")
    private Integer dataSize;
    
    @Column(name = "cost", nullable = false)
    @Builder.Default
    private Integer cost = 1;
    
    @Column(name = "priority", nullable = false)
    @Builder.Default
    private Integer priority = 5;
    
    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;
    
    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        this.createdAt = now;
        this.updatedAt = now;
    }
    
    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }    
    /**
     * 任务状态枚举
     * 对应项目文档中定义的5种状态
     */
    public enum TaskStatus {
        PENDING,    // 提交任务后
        PROCESSING, // 处理中
        SUCCESS,    // 任务成功
        RETRY,      // tasks.status为可重试任务
        FAILED      // 不可重试任务
    }
    
    /**
     * 检查任务是否可以重试
     */
    public boolean canRetry() {
        return this.retryCount < this.maxRetryCount && 
               (this.status == TaskStatus.FAILED || this.status == TaskStatus.RETRY);
    }
    
    /**
     * 增加重试次数
     */
    public void incrementRetryCount() {
        this.retryCount++;
    }
    
    /**
     * 检查任务是否已完成（成功或最终失败）
     */
    public boolean isCompleted() {
        return this.status == TaskStatus.SUCCESS || 
               (this.status == TaskStatus.FAILED && !canRetry());
    }
}
